### PR TITLE
Build C-Ares with Bazel on Windows

### PR DESCRIPTION
**Pull request by rongjiecomputer**
With this PR, gRPC can be built with Bazel on Windows without `--define grpc_no_ares=true` flag.

`config_windows/ares_config.h` is generated by CMake.

### CODE REVIEW

**Comment by rongjiecomputer** [comment]
Ping @apolcyn

**Comment by apolcyn** [comment]
getting to this in the next day

**Comment by apolcyn** [comment]
this mostly lgtm, but one thought is this would have us using the generated `ares_config.h` when building on Windows under bazel, and using the `HAVE_CONFIG_H` ares macro to do so, which would be inconsistent with how we're <i>not</i> using a pre-generated `ares_config.h` header when using the Windows cmake build, though. I'm thinking it would make sense to supply the custom `ares_config.h` files to cmake to match up with this (perhaps in a separate PR though). cc @jtattermusch

**Comment by rongjiecomputer** [comment]
> using the generated `ares_config.h` when building on Windows under bazel, and using the `HAVE_CONFIG_H` ares macro to do so, which would be inconsistent with how we're not using a pre-generated `ares_config.h` header when using the Windows cmake build,

Not sure where the inconsistency comes from. The other 5 `config_*/ares_config.h` files were also generated by CMake. Until C-Ares is updated to a newer version again, CMake build and Bazel build on Windows should be identical (in worst case, only as bad as other 5 platforms).

**Comment by apolcyn** [comment]
> Not sure where the inconsistency comes from. The other 5 config_*/ares_config.h files were also generated by CMake. Until C-Ares is updated to a newer version again, CMake build and Bazel build on Windows should be identical (in worst case, only as bad as other 5 platforms).

Sorry, nevermind that :) you're right

File: third_party/cares/cares.BUILD [hunk 1]
```diff
@@ -138,10 +144,23 @@ cc_library(
     copts = [
         "-D_GNU_SOURCE",
         "-D_HAS_EXCEPTIONS=0",
-        "-DNOMINMAX",
         "-DHAVE_CONFIG_H",
-    ],
+    ] + select({
+        ":windows": [
+            "-DNOMINMAX",
+            "-DWIN32_LEAN_AND_MEAN",
```
Comment (apolcyn) [inline third_party/cares/cares.BUILD]: I might be missing something here, but since the cmake-generated `ares_config.h` has `#undef DWIN32_LEAN_AND_MEAN`, what's the effect of this?

Comment (rongjiecomputer) [reply third_party/cares/cares.BUILD]: > what's the effect of this?

I think they will cancel out each other.

C-Ares add `-DWIN32_LEAN_AND_MEAN` in [`CMakeLists.txt#L169`](https://github.com/c-ares/c-ares/blob/3be1924221e1326df520f8498d704a5c4c8d0cce/CMakeLists.txt#L169), but then undef it unconditionally in [`ares_config.h.cmake#L419`](https://github.com/c-ares/c-ares/blob/3be1924221e1326df520f8498d704a5c4c8d0cce/ares_config.h.cmake#L419).

`ares_config.h` is only included in `ares_setup.h`. After including `ares_config.h`, `ares_setup.h` defines `WIN32_LEAN_AND_MEAN` once again at [`ares_setup.h#L109`](https://github.com/c-ares/c-ares/blob/a5b2e99207fd12d1e9d0413f4a6cdc8d0d9bdda3/ares_setup.h#L109). (No, I have no idea why C-Ares is doing this macro dance).

I think it is best we just follow what `CMakeLists.txt` does and don't think too much.

Comment (apolcyn) [reply third_party/cares/cares.BUILD]: having trouble seeing how this cancels, can this `-DWIN32_LEAN_AND_MEAN` just be removed?

Comment (rongjiecomputer) [reply third_party/cares/cares.BUILD]: I just tested removing this flag and build a project that depends on gRPC. I see no difference. No warnings at all. Should I just remove this flag?

Comment (apolcyn) [reply third_party/cares/cares.BUILD]: Yeah I'd prefer remove it. I believe it has no effect, and it's one less thing to have

Comment (rongjiecomputer) [reply third_party/cares/cares.BUILD]: Done.

File: third_party/cares/cares.BUILD [hunk 2]
```diff
@@ -138,10 +144,22 @@ cc_library(
     copts = [
         "-D_GNU_SOURCE",
         "-D_HAS_EXCEPTIONS=0",
-        "-DNOMINMAX",
         "-DHAVE_CONFIG_H",
-    ],
+    ] + select({
+        ":windows": [
+            "-DNOMINMAX",
```
Comment (apolcyn) [inline third_party/cares/cares.BUILD]: Why is this `NOMINMAX` macro necessary? Trouble seeing why

File: third_party/cares/cares.BUILD [hunk 3]
```diff
@@ -138,10 +144,22 @@ cc_library(
...
+            "-D_CRT_SECURE_NO_DEPRECATE",
```
Comment (apolcyn) [inline third_party/cares/cares.BUILD]: same with these two macros: `_CRT_SECURE_NO_DEPRECATE` and `_CRT_NONSTDC_NO_DEPRECATE` - can you please explain?

Comment (rongjiecomputer) [reply third_party/cares/cares.BUILD]: `NOMINMAX` is to tell `windows.h` not to define `min` and `max` macros. These two macros often cause obscure compile error if you something else with the same name.

`_CRT_SECURE_NO_DEPRECATE` and `_CRT_NONSTDC_NO_DEPRECATE` are needed because MSVC is trying to deprecate perfectly normal C standard functions such as `fopen` (and ask you to use `fopen_s` instead). Even gRPC use these flags: https://github.com/grpc/grpc/blob/master/CMakeLists.txt#L99.

tl;dr: We need these three flags to avoid useless warnings.

Comment (nicolasnoble) [reply third_party/cares/cares.BUILD]: Yeah, these are very standard for Windows building.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```
